SSA/Projects/InstCombine/LLVM/Semantics.lean

```diff
@@ -78,11 +78,25 @@ def add? {w : Nat} (x y : BitVec w) : IntW w :=
 theorem add?_eq : LLVM.add? a b  = .some (BitVec.add a b) := rfl
 
 @[simp_llvm_option]
-def add {w : Nat} (x y : IntW w) : IntW w := do
+def add {w : Nat} (x y : IntW w) (nswnuw : Bool × Bool := ⟨ false , false ⟩ ) : IntW w := do
   let x' ← x
   let y' ← y
-  add? x' y'
-
+  let nsw := nswnuw.1
+  let nuw := nswnuw.2
+  if (nsw ∧  (x'.toInt + y'.toInt) < -(2^(w-1)) ∧ (x'.toInt + y'.toInt) ≥ 2^w) ∨ (nuw ∧  (x'.toNat + y'.toNat) ≥ 2^w) then
```

Comment:
```suggestion
  if (nsw ∧  ((x'.toInt + y'.toInt) < -(2^(w-1)) ∨ (x'.toInt + y'.toInt) ≥ 2^w)) ∨ (nuw ∨  (x'.toNat + y'.toNat) ≥ 2^w) then
```

Pretty sure we need an `∨` instead of an `∨`: It's an overflow if it goes either over or under the range.

Reply:
Even better would be to take this condition and lift it to defs. Something like:

```lean
def intOverflow? (x y : BitVec w) : Prop := 
  ((x.toInt + y.toInt) < -(2^(w-1)) ∨ (x.toInt + y'.toInt) ≥ 2^w))

def natOverflow? (x y : BitVec w) : Prop := (x.toNat + y.toNat) ≥ 2^w
```

And then to write the condition as:

```lean
if (nsw ∧ intOverflow? x' y')  ∨ (nuw ∧ natOverflow? x' y')
```

Reply:
@bollu `intOverflow?` is a `Prop` thus the naming convention demands an UpperCamelCase name. At the risk of bike-shedding, I'd propose `AddSignedWraps` and `AddUnsignedWraps` to mirror the name of the flags (and to explicate this is about whether addition overflows)